lib/spack/spack/cmd/common/arguments.py

```diff
@@ -72,6 +73,35 @@ def _specs(self, **kwargs):
         return sorted(specs.values())
 
 
+class SetParallelJobs(argparse.Action):
+    """Sets the correct value for parallel build jobs.
+
+    The value is pushed on top of the current configuration scopes, and can be
+    retrieved using the spack.config API.
+    """
+    def __call__(self, parser, namespace, values, option_string):
+        # Values is a single integer, type conversion is already applied
+        # see https://docs.python.org/3/library/argparse.html#action-classes
+        if values is not None and values < 1:
+            msg = 'invalid value for argument "{0}" '\
+                  '[expected a positive integer, got "{1}"]'
+            raise ValueError(msg.format(option_string, values))
+
+        # If no option was passed set the default to the number of CPUs,
+        if values is None and not spack.config.get('config:build_jobs'):
+            ncpus = multiprocessing.cpu_count()
+            spack.config.set('config:build_jobs', ncpus, scope='_builtin')
+
+        # Set the number of build jobs according to the argument passed
+        # via the command line
+        if values:
+            spack.config.set('config:build_jobs', values, scope='command_line')
+
+        # Set this namespace attribute for convenience
+        jobs = spack.config.get('config:build_jobs')
+        setattr(namespace, 'jobs', jobs)
```

Comment:
Maybe this is just a matter of preference, but I think the code here could be more clear:

```python
if jobs is None:
    # value was not set on command line
    jobs = spack.config.get('config:build_jobs')
    if jobs is None:
        # there is no default in the config, so set one.
        jobs = multiprocessing.cpu_count()
        spack.config.set('config:build_jobs', jobs, scope='_builtin')
else:
    # jobs was set on the command line; override in command line scope
    spack.config.set('config:build_jobs', value, scope='command_line')

# Set this namespace attribute for convenience
setattr(namespace, 'jobs', jobs)
```

Reply:
On second thought this doesn't need to be nearly as complicated anyway, as we already have a config default here:

https://github.com/spack/spack/blob/develop/lib/spack/spack/config.py#L103

I would remove any setting of values in scope `_builtin` and leave them to `config.py`.

Reply:
The logic above has been simplified a lot. I am resolving the conversation for my own convenience -  please reopen if you think other changes are due.